src/backend/mssql.rs

```diff
@@ -0,0 +1,180 @@
+//! Microsoft SQL Server implementation of a generator
+//!
+//! This module generates strings that are specific to SQL Server
+//! databases. They should be thoroughly tested via unit testing
+
+use super::SqlGenerator;
+use crate::types::{BaseType, Type};
+
+/// A simple macro that will generate a quoted schema prefix if it exists
+macro_rules! quoted_prefix {
+    ($schema:expr) => {
+        $schema
+            .map(|s| format!("[{}].", s))
+            .unwrap_or_else(|| String::new())
+    };
+}
+
+/// A simple macro that will generate a schema prefix if it exists
+macro_rules! prefix {
+    ($schema:expr) => {
+        $schema
+            .map(|s| format!("{}.", s))
+            .unwrap_or_else(|| String::new())
+    };
+}
+
+/// SQL Server generator backend
+pub struct MsSql;
+impl SqlGenerator for MsSql {
+    fn create_table(name: &str, schema: Option<&str>) -> String {
+        format!("CREATE TABLE {}[{}]", quoted_prefix!(schema), name)
+    }
+
+    fn create_table_if_not_exists(name: &str, schema: Option<&str>) -> String {
+        let table = format!("{}[{}]", quoted_prefix!(schema), name);
+
+        format!(
+            "IF NOT EXISTS (SELECT * FROM sysobjects WHERE name='{table}' AND xtype='U') CREATE TABLE {quoted}",
+            table = name,
+            quoted = table,
+        )
+    }
+
+    fn drop_table(name: &str, schema: Option<&str>) -> String {
+        format!("DROP TABLE {}[{}]", quoted_prefix!(schema), name)
+    }
+
+    fn drop_table_if_exists(name: &str, schema: Option<&str>) -> String {
+        format!("DROP TABLE IF EXISTS {}[{}]", quoted_prefix!(schema), name)
+    }
+
+    fn rename_table(old: &str, new: &str, schema: Option<&str>) -> String {
+        let prefix = prefix!(schema);
+        format!(r#"EXEC sp_rename '{}{}', '{}'"#, prefix, old, new)
+    }
+
+    fn alter_table(name: &str, schema: Option<&str>) -> String {
+        format!("ALTER TABLE {}[{}]", quoted_prefix!(schema), name)
+    }
+
+    fn add_column(ex: bool, schema: Option<&str>, name: &str, tt: &Type) -> String {
+        let bt: BaseType = tt.get_inner();
+        use self::BaseType::*;
+
+        #[cfg_attr(rustfmt, rustfmt_skip)] /* This shouldn't be formatted. It's too long */
```

Comment:
Maybe this line could be removed by using a helper function in the `match` block. Something like:

```rust
let f = |ex, name, bt, schema| format!("{}[{}] {}", MsSql::prefix(ex), name, MsSql::print_type(bt, schema));
```

Which allows invocations to become of the form:

```rust
            match bt {
                Text => f(ex, name, bt, schema),
                ...
            }
```